Have createNewBlock() wait for tip, make rpc handle shutdown during long poll and wait methods

Sjors opened the pull request:
This PR prevents Mining interface methods from sometimes crashing when called during startup before a tip is connected. It also makes other improvements like making more RPC methods usable from the GUI. Specifically this PR:

- Adds an `Assume` check to disallow passing negative timeout values to `Mining::waitTipChanged`
- Makes `waitfornewblock`, `waitforblock` and `waitforblockheight` RPC methods usable from the GUI when `-server=1` is not set.
- Changes `Mining::waitTipChanged` to return `optional<BlockRef>` instead of `BlockRef` and return `nullopt` instead of crashing if there is a timeout or if the node is shut down before a tip is connected.
- Changes `Mining::waitTipChanged` to not time out before a tip is connected, so it is convenient and safe to call during startup, and only returns `nullopt` on early shutdowns.
- Changes `Mining::createNewBlock` to block and wait for a tip to be connected if it is called on startup instead of crashing. Also documents that it will return null on early shutdowns.

This allows `waitNext()` (added in https://github.com/bitcoin/bitcoin/pull/31283) to safely assume `TipBlock()` isn't `null`, not even during a scenario of early shutdown.

Finally this PR clarifies long poll behaviour, mostly by adding code comments, but also through an early `break`.

l0rinc commented on src/node/interfaces.cpp:
```diff
@@ -989,6 +989,12 @@ class MinerImpl : public Mining
+        {
+            // Ensure m_tip_block is set so consumers of BlockTemplate can rely on that.
+            WAIT_LOCK(notifications().m_tip_block_mutex, lock);
+            if (!notifications().TipBlock()) return nullptr;
```
If it's not too difficult, can we reproduce this with a test?

Sjors replied:
This depends on the node initialization and shutdown sequence. I'm not sure how to reproduce that in a test. And we might change that sequence in a way that this never happens in the first place.

Sjors commented:
The Stratum v2 Template Provider that I implemented first calls `mining.waitTipChanged(uint256::ZERO)` before calling `mining.createNewBlock()`. This ensures that now matter we order the node & IPC startup sequence, it won't prematurely try to make a block.

See https://github.com/Sjors/bitcoin/pull/49 (`src/sv2/template_provider.cpp` in `Sv2TemplateProvider::ThreadSv2Handler()`).

But other implementers may not realise this, so having `createNewBlock()` return nothing when called too early seems like a good precaution.

Sjors committed:
Have createNewBlock() wait for a tip

- return null on shutdown instead of the last tip
- ignore timeout value node initialization

This allows consumers of BlockTemplate to safely
assume that a tip is connected, instead of having
to account for startup and early shutdown scenarios.

Sjors commented on src/node/interfaces.cpp:
```diff
@@ -989,6 +989,12 @@ class MinerImpl : public Mining
 
     std::unique_ptr<BlockTemplate> createNewBlock(const BlockCreateOptions& options) override
```
The latest push has `waitTipChanged` and `createNewBlock` return null if the node shuts down, and documentation is updated to reflect that.

vasild commented on src/rpc/blockchain.cpp:
```diff
@@ -286,9 +287,13 @@ static RPCHelpMan waitfornewblock()
     Mining& miner = EnsureMining(node);
 
-    auto block{CHECK_NONFATAL(miner.getTip()).value()};
+    BlockRef block{CHECK_NONFATAL(miner.getTip()).value()};
     if (IsRPCRunning()) {
-        block = timeout ? miner.waitTipChanged(block.hash, std::chrono::milliseconds(timeout)) : miner.waitTipChanged(block.hash);
+        std::optional<BlockRef> maybe_block{timeout ? miner.waitTipChanged(block.hash, std::chrono::milliseconds(timeout)) : miner.waitTipChanged(block.hash)};
+        // Return new block unless the node is shutting down
+        if (maybe_block) {
+            block = *CHECK_NONFATAL(maybe_block);
+        }
     }
```
Given that "no tip" is expected here (unlikely but is not a "programming logic error"), maybe avoid `CHECK_NONFATAL()` and use `JSONRPCError`:

```cpp
std::optional<BlockRef> old_block{miner.getTip()};
std::optional<BlockRef> new_block;
if (IsRPCRunning()) {
    const uint256 h{old_block.has_value() ? old_block->hash : uint256::ZERO};
    new_block = timeout > 0 ? miner.waitTipChanged(h, std::chrono::milliseconds(timeout)) : miner.waitTipChanged(h);
}

if (!new_block.has_value()) {
    throw JSONRPCError(RPC_CLIENT_NOT_CONNECTED, "No tip within timeout or shutting down");
}
```

vasild replied:
I mean to avoid `CHECK_NONFATAL()` if the tip is null and use `JSONRPCError` instead.

If the tip is null at the start, then wait for it to be set, if not set within the timeout, then `JSONRPCError`.

Sjors replied:
I ended up with something similar...

vasild commented on src/rpc/blockchain.cpp:
```diff
@@ -335,13 +340,18 @@ static RPCHelpMan waitforblock()
     auto block{CHECK_NONFATAL(miner.getTip()).value()};
     const auto deadline{std::chrono::steady_clock::now() + 1ms * timeout};
     while (IsRPCRunning() && block.hash != hash) {
+        std::optional<BlockRef> maybe_block;
         if (timeout) {
             auto now{std::chrono::steady_clock::now()};
             if (now >= deadline) break;
             const MillisecondsDouble remaining{deadline - now};
-            block = miner.waitTipChanged(block.hash, remaining);
+            maybe_block = miner.waitTipChanged(block.hash, remaining);
         } else {
-            block = miner.waitTipChanged(block.hash);
+            maybe_block = miner.waitTipChanged(block.hash);
+        }
+        // Return new block unless the node is shutting down
+        if (maybe_block) {
+            block = *CHECK_NONFATAL(maybe_block);
         }
     }
```
Similar here:

```cpp
std::optional<BlockRef> block{miner.getTip()};
const auto deadline{std::chrono::steady_clock::now() + 1ms * timeout};
while (IsRPCRunning() && (!block.has_value() || block->hash != hash)) {
    const uint256 h{block.has_value() ? block->hash : uint256::ZERO};
    if (timeout) {
        const auto now{std::chrono::steady_clock::now()};
        if (now >= deadline) break;
        const MillisecondsDouble remaining{deadline - now};
        block = miner.waitTipChanged(h, remaining);
    } else {
        block = miner.waitTipChanged(h);
    }
}

if (!block.has_value()) {
    JSONRPCError(RPC_CLIENT_NOT_CONNECTED, "No tip within timeout or shutting down");
}
```

vasild commented on src/rpc/mining.cpp:
```diff
@@ -801,7 +802,11 @@ static RPCHelpMan getblocktemplate()
+                    throw JSONRPCError(RPC_CLIENT_NOT_CONNECTED, "Shutting down");
+                };
+                tip = CHECK_NONFATAL(maybe_tip)->hash;
                 // Timeout: Check transactions for update
```
Why is comment saying "Timeout: ..." when we could be here if the tip changed (no timeout)? Is the comment wrong?

vasild commented on src/rpc/mining.cpp:
```diff
@@ -801,7 +802,11 @@ static RPCHelpMan getblocktemplate()
-                tip = miner.waitTipChanged(hashWatchedChain, checktxtime).hash;
+                std::optional<BlockRef> maybe_tip{miner.waitTipChanged(hashWatchedChain, checktxtime)};
+                if (!maybe_tip) {
+                    throw JSONRPCError(RPC_CLIENT_NOT_CONNECTED, "Shutting down");
```
Maybe elaborate the message that we could be here if no tip was connected within the timeout: "No tip within timeout or shutting down".

Sjors replied:
I'm introducing a new `RPC_SHUTDOWN_ERROR` which makes sense here. I don't think the message itself matters, since the user knows they shut down the node - and the exact details of where and why the RPC call fails aren't important.

vasild commented on src/rpc/mining.cpp:
```diff
@@ -801,7 +802,11 @@ static RPCHelpMan getblocktemplate()
             while (tip == hashWatchedChain && IsRPCRunning()) {
-                tip = miner.waitTipChanged(hashWatchedChain, checktxtime).hash;
+                std::optional<BlockRef> maybe_tip{miner.waitTipChanged(hashWatchedChain, checktxtime)};
+                if (!maybe_tip) {
+                    throw JSONRPCError(RPC_CLIENT_NOT_CONNECTED, "Shutting down");
+                };
+                tip = CHECK_NONFATAL(maybe_tip)->hash;
```
nit: I think it is redundant to check that `maybe_tip` has value with `CHECK_NONFATAL()` just after `if (!maybe_tip)`. That's like:

```
if (A) {
     throw ...
}
assert(!A);
```

`tip = maybe_tip->hash;` seems fine as well.

Sjors replied:
Indeed, but on the other hand code changes over time, stuff appears between the if statement and the usage of `tip`, then someone changes the if statement, and now we have a crash...

vasild replied:
Ok

vasild commented on src/node/interfaces.cpp:
```diff
@@ -971,24 +971,32 @@ class MinerImpl : public Mining
     std::unique_ptr<BlockTemplate> createNewBlock(const BlockCreateOptions& options) override
     {
+        // Ensure m_tip_block is set so consumers of BlockTemplate can rely on that.
+        if (!waitTipChanged(uint256::ZERO, MillisecondsDouble::max())) return {};
```
The default for the timeout argument is `MillisecondsDouble::max()`, so:

```suggestion
        if (!waitTipChanged(uint256::ZERO)) return {};
```

Sjors replied:
It seems the compiler doesn't know about this default inside the implementation file.

vasild replied:
TIL https://flylib.com/books/en/2.887.1.53/1/

Possible solutions I see:
* drop the default altogether and have all callers use 2 arguments
* use the "non-virtual interface idiom" as described in the chapter above (it is from the book "Effective C++" by Scott Meyers)
* use overload instead of default arguments. Have two methods `waitTipChanged(hash)` and `waitTipChanged(hash, timeout)` (the first could call the second internally).

Sjors replied:
I don't think we should drop the default, because `MillisecondsDouble::max()` is tedious. It seems fine that the interface does it internally, which is simpler than the other two solutions.

Sjors replied:
I could also do this:

```diff
diff --git a/src/node/interfaces.cpp b/src/node/interfaces.cpp
index 66ce64b1c3..b0735d497c 100644
--- a/src/node/interfaces.cpp
+++ b/src/node/interfaces.cpp
@@ -971,7 +971,7 @@ public:
         return BlockRef{tip->GetBlockHash(), tip->nHeight};
     }
 
-    std::optional<BlockRef> waitTipChanged(uint256 current_tip, MillisecondsDouble timeout) override
+    std::optional<BlockRef> waitTipChanged(uint256 current_tip, MillisecondsDouble timeout = MillisecondsDouble::max()) override
     {
         if (timeout > std::chrono::years{100}) timeout = std::chrono::years{100}; // Upper bound to avoid UB in std::chrono
         std::optional<uint256> tip_hash;
@@ -1005,7 +1005,7 @@ public:
     std::unique_ptr<BlockTemplate> createNewBlock(const BlockCreateOptions& options) override
     {
         // Ensure m_tip_block is set so consumers of BlockTemplate can rely on that.
-        if (!waitTipChanged(uint256::ZERO, MillisecondsDouble::max())) return {};
+        if (!waitTipChanged(uint256::ZERO)) return {};
 
         BlockAssembler::Options assemble_options{options};
         ApplyArgsManOptions(*Assert(m_node.args), assemble_options);
```

But that doesn't seem nicer.

Sjors replied:
My takeaway from the article is that having a default for the base class is fine, and that it's better to not use a default for the inherited classes (but it could get ignored).

The patch in my previous comment would be bad. But the current code that calls `waitTipChanged(uint256::ZERO, MillisecondsDouble::max())` should be fine.

ryanofsky replied:
In commit "Have createNewBlock() wait for a tip" (854766b886b6c2ee564b867db64bdbce80064b3b)

Another alternative but probably not better is:

```diff
-        if (!waitTipChanged(uint256::ZERO, MillisecondsDouble::max())) return {};
+        if (!static_cast<Mining*>(this)->waitTipChanged(uint256::ZERO)) return {};
```

I think in longer term after we add an initial waitNext() method and make improvements to it, it will be natural to move more functionality out of Mining class methods into node/miner.cpp or somewhere like that and these methods will get simpler. The node/interfaces.cpp file is mostly meant to contain glue code and wrap other functions, not implement real functionality itself. Since a lot of code needs to wait, I imagine there will be some new waiting utility function this could call instead of waitTipChanged, but using waitTipChanged for now here should be fine.

vasild commented on src/node/interfaces.cpp:
```diff
@@ -971,24 +971,32 @@ class MinerImpl : public Mining
         return BlockRef{tip->GetBlockHash(), tip->nHeight};
     }
```
The message of the first commit has a line that is 211 chars. Would be nice to use the [50/72 rule](https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html) in commit messages.

vasild commented on src/node/interfaces.cpp:
```diff
@@ -971,24 +971,32 @@ class MinerImpl : public Mining
         return BlockRef{tip->GetBlockHash(), tip->nHeight};
```
In commit message of ad3af401c19b8d05ce69011a359db3090b7018e1 `rpc: handle shutdown during long poll and wait methods`:

```diff
- getblocktemplate,  waitfornewblock
+ getblocktemplate, waitfornewblock
```

vasild reviewed (commented):
Approach ACK be15f2e769987b1df2cf1c8f327c5aafe064dbb3

Sjors commented on src/node/interfaces.cpp:
```diff
@@ -971,24 +971,32 @@ class MinerImpl : public Mining
             });
         }
+
+        if (!tip_hash || chainman().m_interrupt) return {};
+
         // Must release m_tip_block_mutex before locking cs_main, to avoid deadlocks.
         LOCK(::cs_main);
-        return BlockRef{chainman().ActiveChain().Tip()->GetBlockHash(), chainman().ActiveChain().Tip()->nHeight};
+        return BlockRef{*Assume(tip_hash), Assume(chainman().ActiveChain().Tip())->nHeight};
```
That seems simpler anyway.

ryanofsky replied:
In commit "rpc: handle shutdown during long poll and wait methods" (ad3af401c19b8d05ce69011a359db3090b7018e1)

Nice suggestion. If this suggestion is applied should also simplify the code above

```diff
-        if (!tip_hash || chainman().m_interrupt) return {};
+        if (chainman().m_interrupt) return {};
```

since `tip_hash` is also relevant. Can also reduce the scope of the `tip_hash` variable.

ryanofsky commented on src/interfaces/mining.h:
```diff
@@ -81,15 +81,21 @@ class Mining
-     * @returns               Hash and height of the current chain tip after this call.
+     *
+     * @retval BlockRef hash and height of the current chain tip after this call.
+     * @retval empty if node is shutting down
```
re: https://github.com/bitcoin/bitcoin/pull/31785#discussion_r1952303154

I don't think suggested text is accurate because if timeout is exceeded while waiting for a new tip, this returns information about the current tip. I don't think that behavior should be changed.

Returning null **only** when node is shutting down should be the simplest and best behavior.

vasild replied:
> if timeout is exceeded while waiting for a new tip, this returns information about the current tip

True. What I meant was that if there is no tip within the timeout. The function does `wait_for(... tip_hash && tip_hash != current_tip ...)` - `tip_hash` to be set and to be different than the current. I meant that it could remain unset while the timeout passes. I see how the comment I suggested is misleading. What about:

```cpp
@retval std::nullopt if the given `timeout` passes before the tip is
connected (there is no tip during startup) or the node is shut down.
```
?

Sjors replied:
I think `@retval std::nullopt if the node is shut down` covers it? Whether shutdown happens before the timeout or exactly at the timeout doesn't matter for the caller.

Sjors replied:
Ah, I think I see the problem. There's a potential scenario where node startup is unusually slow, longer than timeout.

I think it's better in that case to ignore the timeout.

vasild replied:
Yes, this scenario I had in mind.

> in that case to ignore the timeout and instead keep waiting for the node to set a tip

Hmm. What if the tip is not being connected for whatever reason? Wait forever? The current behavior of respecting the timeout looks safer. Would be an odd user experience to provide a timeout and the program to decide to wait longer for whatever reason.

Sjors replied:
See https://github.com/bitcoin/bitcoin/pull/31785#discussion_r1954576364

ryanofsky reviewed (commented):
Code review be15f2e769987b1df2cf1c8f327c5aafe064dbb3. Mostly looks good but I think vasild's suggestion https://github.com/bitcoin/bitcoin/pull/31785#discussion_r1952318566 should be used so waitTipChanged can't return an inconsistent hash and height

Sjors commented:
I changed `waitTipChanged()` to only return `nullopt` during shutdown. This is achieved by extending the `timeout` in the unlikely event that node initialization takes longer.

I introduced a new `RPC_SHUTDOWN_ERROR` code and use it in case `waitfornewblock` and friends are called really early and the node shuts down (probably very hard to reach in practice). See https://github.com/bitcoin/bitcoin/pull/31785#discussion_r1952373326

On question about long polling I still need to investigate: https://github.com/bitcoin/bitcoin/pull/31785#discussion_r1952403604

And this:

> It would be helpful if description said more specifically what previous behavior was.

Sjors commented:
I pushed a commit to clarify long poll behavior.

ryanofsky commented on src/node/interfaces.cpp:
```diff
@@ -971,20 +971,35 @@ class MinerImpl : public Mining
             WAIT_LOCK(notifications().m_tip_block_mutex, lock);
-            notifications().m_tip_block_cv.wait_for(lock, timeout, [&]() EXCLUSIVE_LOCKS_REQUIRED(notifications().m_tip_block_mutex) {
-                // We need to wait for m_tip_block to be set AND for the value
-                // to differ from the current_tip value.
-                return (notifications().TipBlock() && notifications().TipBlock() != current_tip) || chainman().m_interrupt;
-            });
+            do {
+                notifications().m_tip_block_cv.wait_for(lock, timeout, [&]() EXCLUSIVE_LOCKS_REQUIRED(notifications().m_tip_block_mutex) {
+                    // We need to wait for m_tip_block to be set AND for the value
+                    // to differ from the current_tip value.
+                    tip_hash = notifications().TipBlock();
+                    return (tip_hash && tip_hash != current_tip) || chainman().m_interrupt;
+                });
+
+                // In the unlikely event m_tip_block is not set after the timeout,
+                // wait a little longer.
+                timeout = 100ms;
+            } while (!tip_hash && !chainman().m_interrupt);
         }
-        // Must release m_tip_block_mutex before locking cs_main, to avoid deadlocks.
-        LOCK(::cs_main);
-        return BlockRef{chainman().ActiveChain().Tip()->GetBlockHash(), chainman().ActiveChain().Tip()->nHeight};
+
+        if (chainman().m_interrupt) return {};
+
+        // Although we won't use it, since the tip may have changed again
+        // before we know its height, tip_hash should have been set.
+        Assume(tip_hash);
+
+        // Must release m_tip_block_mutex before getTip() locks cs_main, to
+        // avoid deadlocks.
+        return getTip();
     }
```
In commit "rpc: handle shutdown during long poll and wait methods" (fa99eddd926e0d49604d9ca8f21fac85d63652f4)

I think this new behavior of only using `timeout` as time limit for tip changing, not as a time limit for the tip to be set on startup makes a lot of sense and should be much more convenient for callers than previous behavior.

(Previous version in ad3af401c19b8d05ce69011a359db3090b7018e1 would return ambiguous null values on startup that could mean either the node was shutting down or the timeout was reached, and would force callers to immediately make another waitTipChanged call without even giving them any new information in the meantime.)

Only downside of this behavior is that mining API no longer includes a function that can wait with a limited timeout for the tip to be non-null on startup. But if anybody wants this we could add a timeout option to `getTip()` or a new separate `waitForStartup()` method accepting a timeout. Or we could not provide anything, since it is pretty easy for callers to implemented their own timeouts if they are using the capnproto client API.

However, the current implementation of this behavior in fa99eddd926e0d49604d9ca8f21fac85d63652f4 looks a bit janky to me, with double nested wait loop, unnecessary 100ms polling, unnecessary tip_hash variable, and an Assume() check out of place and delayed for no reason. Following is untested, but I think would a simpler implementation that cleans all these issues up:

```c++
    std::optional<BlockRef> waitTipChanged(uint256 current_tip, MillisecondsDouble timeout) override
    {
        if (timeout > std::chrono::years{100}) timeout = std::chrono::years{100}; // Upper bound to avoid UB in std::chrono
        auto deadline{std::chrono::steady_clock::now() + timeout};
        {
            WAIT_LOCK(notifications().m_tip_block_mutex, lock);
            // For callers convenience, wait longer than the provided timeout
            // during startup for the tip to be non-null. That way this function
            // always returns valid tip information when possible and only
            // returns null when shutting down, not when timing out.
            notifications().m_tip_block_cv.wait(lock, [&]() EXCLUSIVE_LOCKS_REQUIRED(notifications().m_tip_block_mutex) {
                return notifications().TipBlock() || chainman().m_interrupt;
            });
            // At this point TipBlock is set, so continue to wait until it is
            // different then `current_tip` provided by caller.
            notifications().m_tip_block_cv.wait_until(lock, deadline, [&]() EXCLUSIVE_LOCKS_REQUIRED(notifications().m_tip_block_mutex) {
                return Assume(notifications().TipBlock()) != current_tip || chainman().m_interrupt;
            });
         }

        if (chainman().m_interrupt) return {};

        // Must release m_tip_block_mutex before getTip() locks cs_main, to
        // avoid deadlocks.
        return getTip();
    }
```

Sjors replied:
Thanks, that indeed looks simpler.

I also think clients should handle timeouts themselves. For `bitcoin-cli` we have `-rpcclienttimeout=0`, other RPC clients presumably have their own timeout mechanisms, and IPC clients can do the same.

Sjors commented on src/rpc/protocol.h:
```diff
@@ -49,6 +49,7 @@ enum RPCErrorCode
     RPC_VERIFY_ALREADY_IN_UTXO_SET  = -27, //!< Transaction already in utxo set
     RPC_IN_WARMUP                   = -28, //!< Client still warming up
     RPC_METHOD_DEPRECATED           = -32, //!< RPC method is deprecated
+    RPC_SHUTDOWN_ERROR              = -37, //!< Node is shutting down
```
First available number.

ryanofsky replied:
> First available number.

Oh, thanks. Didn't notice the numbers were out of order and later ones were used below.

Sjors commented:
I mixed up the commits in my last push, will fix...

ryanofsky commented on src/rpc/blockchain.cpp:
```diff
@@ -286,9 +287,17 @@ static RPCHelpMan waitfornewblock()
     Mining& miner = EnsureMining(node);
 
-    auto block{CHECK_NONFATAL(miner.getTip()).value()};
+    std::optional<BlockRef> old_block{miner.getTip()};
+    if (!old_block) throw JSONRPCError(RPC_SHUTDOWN_ERROR, "Node is shutting down");
+    BlockRef block{*CHECK_NONFATAL(old_block)};
+
+    std::optional<BlockRef> new_block;
     if (IsRPCRunning()) {
-        block = timeout ? miner.waitTipChanged(block.hash, std::chrono::milliseconds(timeout)) : miner.waitTipChanged(block.hash);
+        new_block = timeout ? miner.waitTipChanged(block.hash, std::chrono::milliseconds(timeout)) : miner.waitTipChanged(block.hash);
+        // Return new block unless the node is shutting down
+        if (new_block) {
+            block = *CHECK_NONFATAL(new_block);
+        }
```
In commit "rpc: handle shutdown during long poll and wait methods" (0c06c8108fb43be317d86199750ecd3b7db9a5d2)

I think this code is not quite right because it is assuming that if `getTip()` returns null it means the node is shutting down, when in reality it is starting up. Also the IsRPCRunning() call unnecessarily complicates things and is incorrect as we found previously in #30635. Would also be nice to reduce number of variables used in this code and remove CHECK_NONFATAL calls that can't trigger because they immediately follow null checks. Would suggest:

```diff
--- a/src/rpc/blockchain.cpp
+++ b/src/rpc/blockchain.cpp
@@ -287,18 +287,12 @@ static RPCHelpMan waitfornewblock()
     NodeContext& node = EnsureAnyNodeContext(request.context);
     Mining& miner = EnsureMining(node);
 
-    std::optional<BlockRef> old_block{miner.getTip()};
-    if (!old_block) throw JSONRPCError(RPC_SHUTDOWN_ERROR, "Node is shutting down");
-    BlockRef block{*CHECK_NONFATAL(old_block)};
-
-    std::optional<BlockRef> new_block;
-    if (IsRPCRunning()) {
-        new_block = timeout ? miner.waitTipChanged(block.hash, std::chrono::milliseconds(timeout)) : miner.waitTipChanged(block.hash);
-        // Return new block unless the node is shutting down
-        if (new_block) {
-            block = *CHECK_NONFATAL(new_block);
-        }
-    }
+    std::optional<BlockRef> tip{miner.getTip()};
+    if (!tip) throw JSONRPCError(RPC_IN_WARMUP, "Node is starting up");
+    BlockRef block{*tip};
+    tip = timeout > 0 ? miner.waitTipChanged(block.hash, std::chrono::milliseconds(timeout)) : miner.waitTipChanged(block.hash);
+    if (!tip) throw JSONRPCError(RPC_SHUTDOWN_ERROR, "Node is shutting down");
+    block = *tip;
 
     UniValue ret(UniValue::VOBJ);
     ret.pushKV("hash", block.hash.GetHex());
```

Probably similar simplifications can apply in other methods below.

Sjors replied:
Good point, `getTip()` doesn't wait.

Sjors replied:
I made a variant of this that's a bit more in line with #30635; during startup we'll wait by using the 0 hash.

vasild commented on src/rpc/blockchain.cpp:
```diff
@@ -387,23 +398,31 @@ static RPCHelpMan waitforblockheight()
+    uint256 tip_hash{block ? block->hash : uint256{}};
+    int block_height{block ? block->height : 0};
+
+    while (block_height < height) {
```
Same here, if the user provides `height=0`, and `getTip()` returns `nullopt`, then this will be `while (0 < 0)`, will never enter and `block` will remain `nullopt` and be dereferenced at the end.

vasild commented on src/rpc/blockchain.cpp:
```diff
@@ -286,14 +287,18 @@ static RPCHelpMan waitfornewblock()
+    std::optional<BlockRef> block{miner.getTip()};
+    // If getTip() returns null, which can happen during startup, pass 0 to
+    // waitTipChanged() so it will wait for any tip hash to be set.
+    uint256 tip_hash{block ? block->hash : uint256{}};
```
nit: maybe use `uint256::ZERO`, also used in commit b954af72e4d6a83b3ba701b7ee52d97f2af3254b `Have createNewBlock() wait for a tip`:

```suggestion
    uint256 tip_hash{block ? block->hash : uint256::ZERO};
```

vasild reviewed (commented):
Approach ACK 2042ed105578a71234295b01be39ad8607722abe

Sjors commented on src/rpc/blockchain.cpp:
```diff
@@ -332,22 +337,28 @@ static RPCHelpMan waitforblock()
+    // waitTipChanged() so it will wait for any tip hash to be set.
+    uint256 tip_hash{block ? block->hash : uint256{}};
+
     const auto deadline{std::chrono::steady_clock::now() + 1ms * timeout};
-    while (IsRPCRunning() && block.hash != hash) {
+    while (tip_hash != hash) {
         if (timeout) {
             auto now{std::chrono::steady_clock::now()};
             if (now >= deadline) break;
             const MillisecondsDouble remaining{deadline - now};
-            block = miner.waitTipChanged(block.hash, remaining);
+            block = miner.waitTipChanged(tip_hash, remaining);
         } else {
-            block = miner.waitTipChanged(block.hash);
+            block = miner.waitTipChanged(tip_hash);
         }
+        if (!block) throw JSONRPCError(RPC_SHUTDOWN_ERROR, "Node is shutting down");
+        tip_hash = block->hash;
     }
 
     UniValue ret(UniValue::VOBJ);
-    ret.pushKV("hash", block.hash.GetHex());
-    ret.pushKV("height", block.height);
+    ret.pushKV("hash", block->hash.GetHex());
+    ret.pushKV("height", block->height);
```
I guess it could also crash on a non-existing block hash? That's not unlikely to happen.

Sjors replied:
Added

Sjors commented on src/node/interfaces.cpp:
```diff
@@ -971,24 +971,38 @@ class MinerImpl : public Mining
+            // different then `current_tip` provided by caller.
+            notifications().m_tip_block_cv.wait_until(lock, deadline, [&]() EXCLUSIVE_LOCKS_REQUIRED(notifications().m_tip_block_mutex) {
+                return Assume(notifications().TipBlock()) != current_tip || chainman().m_interrupt;
             });
```
I flipped it.

Sjors replied:
Good point about the comment, I took your suggestion.

Sjors commented:
Addressed @vasild's feedback (and rebased).

ryanofsky reviewed (approved):
Code review ACK 6806651d2f64d51094178cbfb23ef2f4a956aa4d

As mentioned in a comment below, I don't really like how this PR is changing startup behavior of the RPC methods without documenting it, and when it seem like it expands the scope of this PR and makes it harder to understand. Would prefer doing that in #30635 so this PR could be more focused. But current approach does seem ok too.

Sjors commented:
> I think a dedicated PR like https://github.com/bitcoin/bitcoin/pull/30635 would be a better place to update the RPC interface, if that is what we want to do.

I'll rebase the latter PR on top of this, and then reduce the scope here.

vasild reviewed (approved):
ACK 6806651d2f64d51094178cbfb23ef2f4a956aa4d

Would be happy to re-review if suggestions from @ryanofsky are applied, especially https://github.com/bitcoin/bitcoin/pull/31785#discussion_r1960150037

Note wrt RPC changing of the behavior when waiting at startup for the tip to be set -- the current code first sets the tip and then starts the RPC, so this is mostly theoretical or about future changes that might swap the startup order. Thus, IMO, no need for release notes. _Edit:_ I forgot that there is also a change that the RPC will now throw during shutdown. Maybe deserves a release note.

Sjors committed:
rpc: handle shutdown during long poll and wait methods

The waitTipChanged() now returns nullopt if the node is shutting down.

Previously it would return the last known tip during shutdown, but
this creates an ambiguous circumstance in the scenario where the
node is started and quickly shutdown, before notifications().TipBlock()
is set.

The getblocktemplate, waitfornewblock and waitforblockheight RPC
are updated to handle this. Existing behavior is preserved.

Co-authored-by: Ryan Ofsky <ryan@ofsky.org>

Sjors committed:
rpc: clarify longpoll behavior

Move the comparison to hashWatchedChain inside the while loop.

Although this early return prevents the GetTransactionsUpdated()
call in cases where the tip updates, it's only done to improve
readability. The check itself is very cheap (although a more
useful check might not be).

Also add code comments.

Sjors committed:
rpc: drop unneeded IsRPCRunning() guards

This was preventing the (hidden) waitfornewblock, waitforblock and
waitforblockheight methods from being used in the GUI.

The check was added in d6a5dc4a2eaa0d7348804254ca09e75fc3a858ab
when these RPC methods were first introduced.

They could have been dropped when dca923150e3ac10a57c23a7e29e76516d32ec10d
refactored these methods to use waitTipChanged(), which already
checks for shutdown.

Making this change now simplifies the next commit.

Sjors commented on src/rpc/blockchain.cpp:
```diff
@@ -332,22 +337,28 @@ static RPCHelpMan waitforblock()
+    // If getTip() returns null, which can happen during startup, pass 0 to
+    // waitTipChanged() so it will wait for any tip hash to be set.
+    uint256 tip_hash{block ? block->hash : uint256{}};
+
     const auto deadline{std::chrono::steady_clock::now() + 1ms * timeout};
-    while (IsRPCRunning() && block.hash != hash) {
+    while (tip_hash != hash || !block.has_value()) {
```
Agreed, though both disappeared in my last push.

Sjors commented:
> the current code first sets the tip and then starts the RPC, so this is mostly theoretical or about future changes that might swap the startup order.

Indeed, that's what it's supposed to do.

I ended up simplifying things a bit. The original `CHECK_NONFATAL(miner.getTip()` has been replaced with `JSONRPCError(RPC_IN_WARMUP`. So in the unlikely event there's no tip, the user gets a clear error. I also added an `Assume`, so we can fix things if it turns out `SetRPCWarmupFinished` was called too early by our init code.

I restored the original behavior of returning the last tip for both timeout and shutdown. This is possible now because we throw during startup if there's no tip.

I dropped the use of `RPC_SHUTDOWN_ERROR` in `rpc/mining.cpp` (and the new code along with it).

I added a separate commit to drop the `IsRPCRunning` checks. Since these methods are undocumented and hidden, I don't think it needs a release note. That can wait for #30635.

Sjors commented:
I forgot that you can't use `Assume` in RPC code. Using `NONFATAL_UNREACHABLE` instead, though that's not as clear to the user.

Sjors commented:
Four CI machines consistently failing `rpc_blockchain.py` (either `--v1transport` or `--v2transport`)  at:

```py
def assert_waitforheight(height, timeout=2):
   assert_equal(
      node.waitforblockheight(height=height, timeout=timeout)['height'],
      current_height)
```

With "Remote end closed connection without response", but not much useful info otherwise. Can't reproduce locally yet.

Sjors commented:
Replaced `NONFATAL_UNREACHABLE` with `CHECK_NONFATAL` again by avoiding the intermediate `std::optional`, see https://github.com/bitcoin/bitcoin/pull/31785#discussion_r1961627062

Rebased and pushed again, though I haven't figured out the CI issue.

vasild commented:
> Remote end closed connection without response

That probably means `bitcoind` crashed.

>  node0 stderr ... optional ... Assertion 'this->_M_is_engaged()' failed.

Sjors commented:
M_engaged seems to be gcc specific for `std::optional`: https://github.com/gcc-mirror/gcc/blob/master/libstdc%2B%2B-v3/include/std/optional#L291

It presumably happens when you dereference it in a bad way, but then why doesn't it crash when compiled with clang (as on my macOS machine)?

vasild commented:
I would guess the optional is dereferenced when it is nullopt and this is not compiler related. Maybe it only happens on CI due to some timing specifics? CI failures that don't repro locally are the most tedious to nail down. Try adding a lot of `LogError("reached line 123, relevant values %d %s %whatever", ...)` and then inspect the CI log for those.

ryanofsky commented:
This failure is strange. I tried to reproduce the failure from https://cirrus-ci.com/task/6207546920271872?logs=ci#L2923:

```c++
 node0 stderr /usr/bin/../lib/gcc/x86_64-linux-gnu/13/../../../../include/c++/13/optional:479: _Tp &std::_Optional_base_impl<interfaces::BlockRef, std::_Optional_base<interfaces::BlockRef>>::_M_get() [_Tp = interfaces::BlockRef, _Dp = std::_Optional_base<interfaces::BlockRef>]: Assertion 'this->_M_is_engaged()' failed. 
```

using gcc 13.3.0, checking out commit that failed 3872aed9d3cc0c8046272eb5d488ff5ddbe78647 and running the test that failed `test/functional/rpc_blockchain.py --v1transport` but it seemed to pass. I also didn't see any places an optional BlockRef is being referenced in the code reviewing 51aa29ede7bd9a1897f3cf240e70ef180849591d.

Sjors commented:
It's unlikely to be related to starting up the node, because this happens only on the fourth `waitforblockheight` call in the test, where it's waiting for `current_height + 1`. That's the only time when it actually has to wait. It should hit the timeout after 2 milliseconds.

Sjors committed:
Handle negative timeout for waitTipChanged()

Sjors commented:
Perhaps on a slow machine the timeout could go negative here:

```cpp
auto now{std::chrono::steady_clock::now()};
if (now >= deadline) break;
const MillisecondsDouble remaining{deadline - now};
block = miner.waitTipChanged(current_block.hash, remaining);
```

Though that seems unlikely, and I wouldn't expect it to happen consistently. In any case I added better handling for negative timeout.

ryanofsky commented on src/rpc/blockchain.cpp:
```diff
@@ -385,23 +397,29 @@ static RPCHelpMan waitforblockheight()
             const MillisecondsDouble remaining{deadline - now};
-            block = miner.waitTipChanged(block.hash, remaining);
+            if (remaining < 0ms) break;
+            block = miner.waitTipChanged(block->hash, remaining);
```
I reproduced the CI bug and problem happens on this line. Looks like `block->hash` should be replaced by `current_block.hash` here and below

Sjors replied:
Good catch, that's also consistent with `waitforblock`.

ryanofsky commented on src/rpc/blockchain.cpp:
```diff
@@ -385,23 +397,29 @@ static RPCHelpMan waitforblockheight()
-            if (now >= deadline) break;
             const MillisecondsDouble remaining{deadline - now};
-            block = miner.waitTipChanged(block.hash, remaining);
+            if (remaining < 0ms) break;
```
In commit "rpc: handle shutdown during long poll and wait methods" (edfb4ad204e71fc266be55748290340b8791e12f)

This change seems equivalent to previous code, and previous code seems clearer. This change also makes waitforblockheight implementation less consistent with waitforblock. Would suggest:

```diff
--- a/src/rpc/blockchain.cpp
+++ b/src/rpc/blockchain.cpp
@@ -406,8 +406,8 @@ static RPCHelpMan waitforblockheight()
         std::optional<BlockRef> block;
         if (timeout) {
             auto now{std::chrono::steady_clock::now()};
+            if (now >= deadline) break;
             const MillisecondsDouble remaining{deadline - now};
-            if (remaining < 0ms) break;
             block = miner.waitTipChanged(current_block.hash, remaining);
         } else {
             block = miner.waitTipChanged(current_block.hash);

```

Sjors replied:
Indeed, I was briefly confused about this, but there's no need to change it.

ryanofsky commented on src/interfaces/mining.h:
```diff
@@ -89,10 +89,13 @@ class Mining
      * @param[in] options options for creating the block
-     * @returns a block template
+     * @retval BlockTemplate a block template.
+     * @retval std::nullopt if the node is shut down.
```
In commit "Have createNewBlock() wait for a tip" (a5b34064891814a998cac73c32f60d55a93e6811)

This is actually returning nullptr not nullopt

ryanofsky reviewed (approved):
Code review ACK b74b6c95eeba57ee6f627ceae0deeb6b14eadc0c

ryanofsky commented:
From the PR description:

> The first commit allows these three methods to be used in the GUI, by removing the `IsRPCRunning()`. This simplifies the other changes.

This currently happens in the second commit, not the first commit.

I think in general PR description is a little hard to follow though. Here is how I would describe this PR:

- This PR prevents `Mining` interface methods from sometimes crashing when called during startup before a tip is connected. It also makes some RPC methods usable from the GUI. Specifically this PR:
  - Adds an `Assume` check to disallow passing negative timeout values to `Mining::waitTipChanged`
  - Makes `waitfornewblock` `waitforblock` and `waitforblockheight` functions usable from the GUI when `-server=1` is not set.
  - Changes `Mining::waitTipChanged` to return `optional<BlockRef>` instead of `BlockRef` and return `nullopt` instead of crashing if there is a timeout or if the node is shut down before a tip is connected.
  - Changes `Mining::waitTipChanged` to never timeout before a tip is connected, so it is convenient and safe to call during startup, and only returns `nullopt` on early shutdowns.
  - Changes `Mining::createNewBlock` to block and wait for a tip to be connected if it is called on startup instead of crashing. Documents that it will return null on early shutdowns.

Sjors commented on src/rpc/blockchain.cpp:
```diff
@@ -286,12 +287,17 @@ static RPCHelpMan waitfornewblock()
-    auto block{CHECK_NONFATAL(miner.getTip()).value()};
-    block = timeout ? miner.waitTipChanged(block.hash, std::chrono::milliseconds(timeout)) : miner.waitTipChanged(block.hash);
+    // Abort if RPC came out of warmup too early
+    BlockRef current_block{CHECK_NONFATAL(miner.getTip()).value()};
```
Yes, I'll update the commit message.

Sjors commented:
Addressed feedback and updated the PR description.

vasild reviewed (approved):
ACK 9725cd6a5c20302da3dcae8fc1156458537ce1df

ryanofsky reviewed (approved):
Code review ACK 9725cd6a5c20302da3dcae8fc1156458537ce1df. Just minor suggestions implemented since last review: changing comments and swapping current_block / block in waitfornewblock.

Sjors commented:
Rebased after #31283. There was no conflict, but the PR description makes reference to `waitNext()` so this should make it easier to (re-)review with that context in mind.

ismaelsadeeq reviewed (commented):
Concept ACK

In 64a2795fd4fe223a55564c31e9fa36264e79ac22   "rpc: handle shutdown during long poll and wait methods"

is there a use for returning the last known tip?

If not, should we simplify the behavior by always returning nullopt when the tip has not changed or no better block is available?

Or It will be more consistent to return the last known tip if it is set, when this edge case happen handle it by returning nullopt.

vasild reviewed (approved):
ACK cc1001f3bf17b31512c05fb359e09483a07fb2a3

ryanofsky commented on src/interfaces/mining.h:
```diff
@@ -89,15 +89,16 @@ class Mining
     /**
      * Waits for the connected tip to change. During node initialization, this will
-     * wait until the tip is connected.
+     * wait until the tip is connected (regardless of `timeout`).
```
In commit "Have createNewBlock() wait for a tip" (db14ca3556ca792546bf4343feb733271333690f)

Commit message is a little confusing because it doesn't mention the timeout change. Would be clearer if it said the commit was changing two things (1) returning null on shutdown instead of last tip (2) ignoring timeout value during startup instead of returning 0 if timeout elapsed before tip was connected

Sjors replied:
I improved the commit message.

Sjors replied:
Will do if I need to retouch.

ryanofsky reviewed (approved):
Code review ACK cc1001f3bf17b31512c05fb359e09483a07fb2a3. Only change since last review was rebasing after #31283

re: https://github.com/bitcoin/bitcoin/pull/31785#pullrequestreview-2702397335

> is a use for returning the last known tip?

I think the main use for returning the last known tip when the tip hasn't changed is just to allow the caller distinguish between the case where the tip has not changed from the case where the node is shutting down. If it returned null in both cases we would need to have to add a separate isShuttingDown method to distinguish them and code calling waitTipChanged in a loop would be more complicated.

I don't think there is a use for returning the last known tip when the known is shutting down, just because caller can assume the tip at that point is the same value they passed as the `current_tip` argument.

ryanofsky reviewed (approved):
Code review ACK 05117e6e17f9a2d9a18a5b32570808c8907febb3, just updated a commit message since last review

sedited reviewed (approved):
ACK 05117e6e17f9a2d9a18a5b32570808c8907febb3

Sjors commented on src/node/interfaces.cpp:
```diff
@@ -1070,22 +1070,34 @@ class MinerImpl : public Mining
+
+        if (chainman().m_interrupt) return {};
+
+        // Must release m_tip_block_mutex before getTip() locks cs_main, to
```
Will look into moving it if I need to retouch more.

vasild reviewed (approved):
ACK 05117e6e17f9a2d9a18a5b32570808c8907febb3

achow101 commented:
ACK 05117e6e17f9a2d9a18a5b32570808c8907febb3
